### PR TITLE
PERF Speed up adapter injection by avoiding repeated adapter-prefix scans

### PR DESCRIPTION
**Pull request by eSVeeF**
### What does this PR do?

When adding a new adapter to a model that already contains adapters, `BaseTuner.inject_adapter` needs to skip the internals of existing tuner layers.

The previous implementation checked every module name against every existing adapter prefix using repeated `startswith` calls. This becomes increasingly expensive as model depth and the number of adapter modules grow.

This PR replaces those string-prefix scans with identity-based membership:

- Collect the identities of descendants inside existing `BaseTunerLayer` instances
- Skip those descendants during injection
- Keep the `BaseTunerLayer` itself eligible so the new adapter can be added to it

The behavior is unchanged, but the filtering step scales much better with model depth.

### Benchmark

https://gist.github.com/eSVeeF/2379ebb60841544520d2d0d7bb3b4554

The benchmark uses a synthetic CPU-only transformer-like model with one PyTorch thread. It compares the old prefix-scan implementation with the new identity-based implementation.

Results across three runs:

| Model depth | Modules | Filtering speedup |
| ---: | ---: | ---: |
| 32 | 802 | 3.8x-4.3x |
| 64 | 1,602 | 7.5x-7.7x |
| 128 | 3,202 | 9.7x-14.7x |
| 256 | 6,402 | 25.0x-33.7x |

As an additional context (rather than a direct before/after comparison), the end-to-end measurement includes model construction and first-adapter injection:

| Model depth | Current second-adapter injection |
| ---: | ---: |
| 32 | 54-62 ms |
| 64 | 115-126 ms |
| 128 | 342-381 ms |
| 256 | 612-699 ms |

### Validation

Ran:

```bash
pytest -q --no-cov tests/test_tuners_utils.py
```

Result:

```text
219 passed, 16 skipped
```

Also ran the focused low-level API tests:

```bash
pytest -q --no-cov tests/test_low_level_api.py
```

Result:

```text
1 failed, 50 passed, 1 xfailed
```

The failure is unrelated to this change. The test
`test_inject_from_state_dict_model_with_weight_conversion` expects the loaded `CLIPTextModel` to expose `model.encoder`, but the installed Transformers version exposes the encoder under `model.text_model`. The failure occurs in the test's sanity-check assertion before it validates the adapter injection result.

### CODE REVIEW

File: src/peft/tuners/tuners_utils.py [hunk 1]
```diff
@@ -868,10 +868,14 @@ def inject_adapter(
         # MATCHING & CREATING MODULES #
         ###############################
 
-        existing_adapter_prefixes = []
-        for key, module in named_modules:
-            if isinstance(module, BaseTunerLayer):
-                existing_adapter_prefixes.append(key + ".")
+        # Skip existing tuner internals, but keep the tuner layer itself eligible for adding the new adapter
+        existing_adapter_children = {
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py]: I think in this case, an explicit loop helps with readability.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for this PR to speed up adding new adapters. I think we have a few "accidentally quadratic" (or even cubic) issues when it comes to matching keys. Normally, those are not big issues, as the time it takes to add new modules is still relatively small in the grand scheme of things, but it doesn't mean we can't add improvements.

The change you provide looks correct, I checked that the results are identical. I just take some issue with the benchmark, which is very synthetic. Did you check how this affects realistic use cases? E.g. take a ` "meta-llama/Llama-3.2-3B"` and check the runtime there. If the speed of adding a new adapter in that case is dominated by other factors, this optimization may not be needed.

Regarding the benchmark script itself: I don't think it's necessary that it checks old and new behavior in one go. A simple script that measures the times is enough, we can switch the branches and re-run it for comparison.

**Comment by eSVeeF** [comment]
Thanks for checking this.

I updated and reran the benchmark using `meta-llama/Llama-3.2-3B` and `Qwen/Qwen3-0.6B`, measuring only the time to add a second adapter after model loading and first-adapter injection. But the results do not show any speedup with the real models. 

Across three executions of the benchmark for both models with `--warmup-runs 1 --benchmark-runs 10`:

- `Qwen/Qwen3-0.6B`: main branch `28.8-39.8 ms`, this branch `32.3-35.3 ms`
- `meta-llama/Llama-3.2-3B`: main branch `35.2-42.5 ms`, this branch `33.5-42.7 ms`

Even though the matching algorithm avoids repeated prefix scans, the times are similar.

I can rename the PR since I don’t want to make its practical impact sound bigger than it is.

I’m also happy to close the PR. WDYT?

**Comment by BenjaminBossan** [comment]
Thanks for doing the new benchmark. This means that the bottleneck is elsewhere and the saved time has no significant impact on the real world usage. My proposal would thus be to close the PR, we can revisit it if a user runs into this issue in their use case.

**Comment by eSVeeF** [comment]
Sounds good, happy to revisit if needed
